### PR TITLE
feat(dsh): upgrade to dsh 0.1.5-rc.2 and spill into the workspace

### PR DESCRIPTION
**Pull request by zechengz**
## Summary

Upgrades `@struktoai/mirage-dsh` to dsh `0.1.5-rc.2` (cordis `^4.0.2`) and closes a gap where an oversized tool result was written somewhere the agent cannot read.

- **Version bump.** The dsh family's npm `latest` tag is stale (`dsh-fs@latest` is still `0.0.1-rc.1`); the live train is `next`. Bumping `packages/dsh/package.json` alone does nothing, because a `pnpm.overrides` block in `typescript/package.json` pins 13 `@deepseek-ai/dsh-*` packages, so that block moves too. The same override had been silently rewriting stale specs in `examples/typescript/package.json` (`0.0.1-rc.1` / `0.0.1-rc.5`); those are now honest.
- **`readByteRange`** is a new abstract method on `FileSystem`, so the build fails without it. It maps onto `ops.readFile(key, {offset, size})`, which already normalizes the two edge cases the seam specifies: a zero-length window short-circuits before any backend call, and a store's 416 becomes the empty POSIX answer.
- **`processPathFromHostPath`** maps a host path into the workspace through a disk mount, the one place the two worlds hold the same file. Every other mount declines, which is the honest answer for an S3 key or a Slack message.
- **`MirageSpillStore` on `ctx.spillStore`.** `spill-policy` is live in dsh-base at `maxInlineBytes: 50000`: it replaces any tool result over ~50 KB with a preview plus a locator and tells the model "Use read with offset/limit, or grep this path". dsh-base stores that through `spill-local`, which writes to the host tmpdir, so in a mirage world the model's follow-up grep hits a path that does not exist and an oversized result degrades to truncation with a dead pointer. The artifact now goes through the same op door as every other effect, so the locator is an ordinary workspace path. A write failure rejects, per the seam contract, and the policy then keeps the inline result, which beats a dead locator.

### On `ctx.subprocess`

dsh's execution-world story has moved: `dsh-e2b` is now three packages and provides `ctx.subprocess` rather than `ctx.shell`, with `bash-sandbox`, `terminal-bash` and `lsp-stdio` sitting on top. This PR deliberately does not follow that yet. Audited exposure: exactly three dsh-base rows inject `subprocess` (`bash-sandbox`, `pwsh-sandbox`, `tool-fs-search`) and the patch disables all three, so the row is inert in dsh-base today. The hole only opens if a profile adds a subprocess consumer of its own. That is written up in `cordis.patch.yml` and is the natural next change.

## Test plan

- `packages/dsh` 229 tests (up from ~200): `readByteRange` window/EOF/zero-length/binary/directory cases, `processPathFromHostPath` mapping and declines, the spill store's naming, session scoping and rejection path, plus an integration test that spills a result and then recovers a line from it with `grep` through `ctx.shell`.
- Full TS suite green: core 12455, node 2994, browser 311, agents 199, server 207, cli 96, opencode 2, zero failures.
- `tsc --noEmit`, eslint, knip, prettier, `pnpm -r build`, and the examples typecheck against the bumped versions.
- `pre-commit run --all-files` exit 0, plus the non-hook CI gates: layout parity (234 = baseline), spec parity, config door parity, barrel surface, docs frontmatter, case targets.
- `uv run pytest` exit 0 (no Python files in the diff; mypy also covers all 2136 Python sources).

### CODE REVIEW

File: typescript/packages/dsh/src/fs.ts [hunk 1]
```diff
@@ -243,6 +244,42 @@ export class MirageFileSystem extends FileSystem {
...
+      if (!(resource instanceof DiskResource)) continue
+      const rel = relative(resource.root, host)
+      // A path outside the root escapes through `..` or answers absolute
+      // (a different drive on win32); neither is under this mount.
+      if (rel.startsWith('..') || isAbsolute(rel)) continue
+      // `prefix` always carries a trailing slash, and `rel` is empty for
+      // the root itself, so the join is a concatenation and the slash is
+      // trimmed back off unless the mount is the workspace root.
+      const virtual = `${entry.prefix}${rel.split(sep).join('/')}`
+      return virtual.length > 1 && virtual.endsWith('/') ? virtual.slice(0, -1) : virtual
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/dsh/src/fs.ts]: Fixed in 2f1f94066. Confirmed real: a disk at `/work` with a RAM mount at `/work/cache` mapped `<root>/cache/x.txt` to `/work/cache/x.txt`, which dispatch routes to the RAM child, so the returned path named another resource's bytes. That is the confidently-wrong answer this seam is supposed to avoid.

Added `ownerPrefixOf`, which resolves a virtual path to the deepest mount whose prefix covers it (the rule dispatch itself follows). A computed path whose owner is not the mount it came from is skipped and the scan continues, so another disk mount can still claim it.

Regression test `declines a path shadowed by a mount nested under the disk mount` in `fs.test.ts`, which also asserts the rest of the disk mount still maps. Verified it fails on the pre-fix code.

File: typescript/packages/dsh/src/fs.ts [hunk 2]
```diff
@@ -243,6 +244,42 @@ export class MirageFileSystem extends FileSystem {
...
+    const workspace = this.ctx.mirage.workspaceIfReady
+    if (workspace === null) return undefined
+    const host = resolve(hostPath)
+    for (const entry of workspace.mounts()) {
+      const { resource } = entry
+      if (!(resource instanceof DiskResource)) continue
+      const rel = relative(resource.root, host)
+      // A path outside the root escapes through `..` or answers absolute
+      // (a different drive on win32); neither is under this mount.
+      if (rel.startsWith('..') || isAbsolute(rel)) continue
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/dsh/src/fs.ts]: Fixed in 2f1f94066. Correct: `relative()` answers `..draft/a.txt` for an ordinary in-root file, and the prefix test read that as an escape.

The check now lives in `escapesRoot` and is `rel === ".." || rel.startsWith(`..${sep}`) || isAbsolute(rel)`, so only a bare `..` or a leading `..` segment counts (the absolute arm stays for the win32 different-drive case, where there is no relative route at all).

Regression test `maps an in-root name that merely begins with two dots` in `fs.test.ts`, which pins both halves: `<root>/..draft/a.txt` maps, and `<root>/..` still declines. Verified it fails on the pre-fix code.

File: typescript/packages/dsh/src/fs.ts [hunk 3]
```diff
@@ -243,6 +282,47 @@ export class MirageFileSystem extends FileSystem {
...
+      // the root itself, so the join is a concatenation and the slash is
+      // trimmed back off unless the mount is the workspace root.
+      const joined = `${entry.prefix}${rel.split(sep).join('/')}`
+      const virtual = joined.length > 1 && joined.endsWith('/') ? joined.slice(0, -1) : joined
+      // A mount nested under this one owns its own subtree, and dispatch
+      // routes the path there, so the disk file at this host location is
+      // not what the virtual path reads. Keep looking rather than name a
+      // path that answers with another resource's bytes.
+      if (ownerPrefixOf(virtual, mounts) !== entry.prefix) continue
+      return virtual
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/dsh/src/fs.ts:348]: Fixed in adbad55db. Right, and it is the same class as the child-mount finding one plane over: I checked the mount table but not the link table.

Added `shadowedByLink`, which declines when `links.follow(virtual) !== virtual`. `follow` resolves link components including the final one, so it covers both the exact-name case you describe and an ancestor link. A chain it refuses to resolve (a cycle) is declined too rather than raised, since this is a lookup and the honest answer is then "no virtual path".

Two regression tests in `fs.test.ts`, both verified red on the pre-fix code:
- `declines a path a namespace symlink shadows`, using your exact ordering (link created first, disk grows the file afterwards by an external write that bypasses mirage), and asserting a sibling the link does not cover still maps.
- `declines a path whose ancestor is a namespace symlink`.

File: typescript/packages/dsh/src/spill-store.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  private readonly dir: string
+
+  constructor(ctx: Context, config: MirageSpillConfig = {}) {
+    super(ctx)
+    this.dir = config.dir ?? DEFAULT_SPILL_DIR
+  }
+
+  async saveText(input: SaveTextSpill): Promise<SpillRef> {
+    const workspace = await this.ctx.mirage.ready
+    const dir = `${this.dir}/${sessionDirName(input.owner.sessionId)}`
```
[the remark this answers is not in the document]
Comment (zechengz) [reply typescript/packages/dsh/src/spill-store.ts]: Fixed in 1c4ee69fb. Confirmed by probe, and it is worse than described: BOTH spellings were dead. With `dir: /tmp/artifacts/../spill` on the RAM mount, `cat` on the literal locator exits 1 and `cat` on its normalized spelling exits 1 with `No such file or directory`. The op door writes the `..` literally and every reader normalizes it away, so the bytes land on a key nothing can address. Exactly the dead locator this store exists to prevent, in the store itself.

`config.dir` is now `posix.resolve("/", dir)` once at construction, and a relative dir is refused there rather than silently rooted. The per-save joins use `posix.join`, so the `/` and `//` edge cases cannot come back.

Four regression tests in `spill-store.test.ts`: the `..` case asserting the locator is canonical AND that a shell `cat` on it returns the payload (the end-to-end readback, which is the property that actually matters), `//./` collapsing, a dir climbing past the root being clamped, and the relative-dir refusal. All four were seen failing before the fix landed.

File: docs/typescript/agents/dsh.mdx [hunk 5]
```diff
@@ -9,7 +9,7 @@ icon: /images/deepseek-logo.svg
 ## Install
 
 ```bash
-npm install @struktoai/mirage-dsh @struktoai/mirage-node @deepseek-ai/cordis @deepseek-ai/dsh-fs @deepseek-ai/dsh-shell
+npm install @struktoai/mirage-dsh @struktoai/mirage-node @deepseek-ai/cordis @deepseek-ai/dsh-fs @deepseek-ai/dsh-shell @deepseek-ai/dsh-spill
```
[the remark this answers is not in the document]
Comment (zechengz) [reply docs/typescript/agents/dsh.mdx:12]: Fixed in 1c4ee69fb. Correct, and a real gap: I updated the npm install line and the bundle-patch paragraph but left the programmatic compose example registering only `MirageService`, `MirageFileSystem` and `MirageShellExecutor`. Anyone following that path installed the package and kept dsh's host-local spill store, which is the precise failure this PR is about.

`MirageSpillStore` is now in the import and registered on the context, and the surrounding line says "Four Cordis plugins" rather than three. Added a short paragraph under the example saying what the plugin buys and that it defaults to `/tmp/dsh-spill`, so a reader whose world has no writable `/tmp` knows to pass `dir`.

Note the bundle path was already covered: `cordis.patch.yml` disables `spill-local` and inserts `mirage-spill`, so `dsh plugin add` users were fine. It was only the hand-composed path that was wrong.
